src/main/java/springboot/rest/specifications/CustomSpecifications.java

```diff
@@ -199,28 +199,58 @@ public Predicate searchInAllAttributesPredicate(CriteriaBuilder builder, Root ro
 
     }
 
+    /**
+     * creates an equality predicate.
+     * Basically it checks for type and does the right predicate.
+     * @param builder CriteriaBuilder
+     * @param root Root
+     * @param join Join Context
+     * @param a Attribute
+     * @param val Object for comparison (from user)
+     * @return Predicate
+     */
     private Predicate createEqualityPredicate(CriteriaBuilder builder, Root root, Join join, Attribute a, Object val) {
         if (isNull(a, val)) {
             if (a.isAssociation() && a.isCollection()) {
-                return builder.isEmpty(root.get(a.getName()));
+                return builder.isEmpty(join != null ? join.get(a.getName()) : root.get(a.getName()));
             }
             else if(isPrimitive(a)) {
-                return builder.isNull(root.get(a.getName()));
+                return builder.isNull(join != null ? join.get(a.getName()) : root.get(a.getName()));
             }
             else {
-                return root.get(a.getName()).isNull();
+                return join != null ? join.get(a.getName()).isNull() : root.get(a.getName()).isNull();
             }
-        } else if (isEnum(a)) {
-            return builder.equal(root.get(a.getName()), Enum.valueOf(Class.class.cast(a.getJavaType()), (String) val));
-        } else if (isPrimitive(a)) {
-            return builder.equal(root.get(a.getName()), val);
-        } else if (a.isAssociation()) {
-            return prepareJoinAssociatedPredicate(root, a, val);
-        } else if(isSerializable(a)) {
-            try {
-                return builder.equal (root.get(a.getName()), a.getJavaType().getConstructors()[0].newInstance(val));
-            } catch (InstantiationException | IllegalAccessException | InvocationTargetException e) {
-                throw new IllegalArgumentException("equality/inequality is currently supported on primitives, enums and serializables with string constructor", e);
+        }
+        else if (join == null) {
```

Comment:
this part is really complicated after the copy paste for join  == null or join != null